File: numbers/src/main/java/org/dicio/numbers/formatter/datetime/FormatStringCollection.java
package org.dicio.numbers.formatter.datetime;

import com.grack.nanojson.JsonObject;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class FormatStringCollection {

    private static class PatternFormatPair {
        final Pattern pattern;
        final FormatString formatString;

        private PatternFormatPair(final JsonObject jsonObject) {
            pattern = Pattern.compile(jsonObject.getString("match"));
            formatString = new FormatString(jsonObject.getString("format"));
        }
    }

    private final FormatString defaultFormat;
    private final List<PatternFormatPair> patternFormats;

    public FormatStringCollection(final JsonObject jsonObject) {
        defaultFormat = new FormatString(jsonObject.getString("default"));
        patternFormats = new ArrayList<>();
        for (int i = 1; jsonObject.has(String.valueOf(i)); ++i) {
            patternFormats.add(new PatternFormatPair(jsonObject.getObject(String.valueOf(i))));
        }
    }

    public FormatString getMostSuitableFormatString(final int number) {
        final String numberString = String.valueOf(number);
        for (final PatternFormatPair patternFormat : patternFormats) {
            if (patternFormat.pattern.matcher(numberString).matches()) {
                return patternFormat.formatString;
            }
        }
        return defaultFormat;
    }
}

File: numbers/src/main/java/org/dicio/numbers/formatter/datetime/FormatStringCollection.kt
package org.dicio.numbers.formatter.datetime

import com.grack.nanojson.JsonObject
import java.util.regex.Pattern

class FormatStringCollection(jsonObject: JsonObject) {
    private class PatternFormatPair(jsonObject: JsonObject) {
        val pattern: Pattern = Pattern.compile(jsonObject.getString("match"))
        val formatString: FormatString = FormatString(jsonObject.getString("format"))
    }

    private val defaultFormat = FormatString(jsonObject.getString("default"))
    private val patternFormats: MutableList<PatternFormatPair> = ArrayList()

    init {
        var i = 1
        while (jsonObject.has(i.toString())) {
            patternFormats.add(PatternFormatPair(jsonObject.getObject(i.toString())))
            ++i
        }
    }

    fun getMostSuitableFormatString(number: Int): FormatString {
        val numberString = number.toString()
        for (patternFormat in patternFormats) {
            if (patternFormat.pattern.matcher(numberString).matches()) {
                return patternFormat.formatString
            }
        }
        return defaultFormat
    }
}
